File: infrastructures/rest-ddd/java/src/main/java/com/example/restddd/application/mappers/IMapper.java
/**
 * Base Mapper Interface
 * ALL transformations MUST use explicit mappers.
 */

package com.example.restddd.application.mappers;

public interface IMapper<TDto, TModel, TWriteEntity, TReadEntity, TResponse> {
    /**
     * Interface for mapping between layers.
     * 
     * Rules:
     * - MANDATORY for all transformations
     * - DTO ↔ BMO ↔ Entity conversions
     * - Located in /application/mappers/
     */

    /**
     * Map Request DTO → BMO (Business Model Object).
     * Used by controllers when receiving requests.
     */
    TModel toModelFromRequest(TDto p_dto);

    /**
     * Map BMO → WriteEntity.
     * Used by command repositories internally.
     */
    TWriteEntity toWriteEntity(TModel p_model);

    /**
     * Map WriteEntity → BMO.
     * Used by command repositories internally.
     */
    TModel toModelFromWriteEntity(TWriteEntity p_entity);

    /**
     * Map ReadEntity → Response DTO.
     * Used by query operations.
     */
    TResponse toResponseFromReadEntity(TReadEntity p_entity);

    /**
     * Map BMO → Response DTO.
     * Used after command operations.
     */
    TResponse toResponseFromModel(TModel p_model);
}

File: infrastructures/ddd/csharp/src/Application/Mappers/IMapper.cs
namespace OddlyDdd.Application.Mappers
{
    /// <summary>
    /// Base interface for mappers that transform between different object types.
    /// ALL transformations between DTOs, BMOs, and Entities MUST use explicit mappers.
    /// Mappers are MANDATORY for the following transformations:
    /// - DTO → BMO (Request to Business Model)
    /// - BMO → WriteEntity (Business Model to Persistence)
    /// - WriteEntity → BMO (Persistence to Business Model)
    /// - ReadEntity → Response DTO (Query Result to Response)
    /// - BMO → Response DTO (Business Model to Response)
    /// </summary>
    /// <typeparam name="TRequestDto">The request DTO type</typeparam>
    /// <typeparam name="TResponseDto">The response DTO type</typeparam>
    /// <typeparam name="TModel">The business model (BMO) type</typeparam>
    /// <typeparam name="TWriteEntity">The write entity type</typeparam>
    /// <typeparam name="TReadEntity">The read entity type</typeparam>
    public interface IMapper<TRequestDto, TResponseDto, TModel, TWriteEntity, TReadEntity>
    {
        /// <summary>
        /// Maps a request DTO to a business model.
        /// Used when receiving data from the API layer.
        /// </summary>
        TModel ToModelFromRequest(TRequestDto p_dto);

        /// <summary>
        /// Maps a business model to a write entity.
        /// Used when persisting data to the database (command side).
        /// </summary>
        TWriteEntity ToWriteEntity(TModel p_model);

        /// <summary>
        /// Maps a write entity to a business model.
        /// Used when loading data from the database for business logic.
        /// </summary>
        TModel ToModelFromWriteEntity(TWriteEntity p_entity);

        /// <summary>
        /// Maps a read entity to a response DTO.
        /// Used for query operations - bypasses BMO in read path for performance.
        /// </summary>
        TResponseDto ToResponseFromReadEntity(TReadEntity p_entity);

        /// <summary>
        /// Maps a business model to a response DTO.
        /// Used when returning data from command operations.
        /// </summary>
        TResponseDto ToResponseFromModel(TModel p_model);
    }
}
